test/core.jl

```diff
@@ -4452,3 +4452,15 @@ end
 #end
 #@test local_innersig(Int32(2)) == ((Int32(2), Int32(1), Int32(2)im), (Int32(2), UInt32(1)))
 #@test local_innersig(Int64(3)) == ((Int64(3), Int64(1), Int64(3)im), (Int64(3), UInt64(1)))
+
+# issue #17572
+function f17572{A}(::Type{Val{A}})
+    return Tuple{Int}(Tuple{A}((1,)))
+end
+try
+    f17572(Val{0})
+    @test false
+catch me
+    # cannot convert to non-type `0`
+    @test isa(me,MethodError) && me.f === convert
+end
```

Comment:
this test doesn't fail for me on master. I think you need to call `code_typed` directly and show that there is no error being thrown.

also, since it's not testing core, it should go in `test/inference.jl` (so you don't get rebase conflicts :)